subsys/zbus/zbus.c

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright (c) 2022 Rodrigo Peixoto <rodrigopex@gmail.com>
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/printk.h>
+#include <zephyr/zbus/zbus.h>
+LOG_MODULE_REGISTER(zbus, CONFIG_ZBUS_LOG_LEVEL);
+
+#if defined(CONFIG_THREAD_NAME)
+#define _ZBUS_CURRENT_THREAD_INFO k_thread_name_get(k_current_get()), k_current_get()
+#else
+#define _ZBUS_CURRENT_THREAD_INFO "_", k_current_get()
+#endif /* CONFIG_THREAD_NAME */
+
+inline k_timeout_t _zbus_timeout_remainder(uint64_t end_ticks)
+{
+	int64_t now_ticks = sys_clock_tick_get();
+
+	return K_TICKS((k_ticks_t)MAX(end_ticks - now_ticks, 0));
+}
+
+int zbus_obs_set_enable(struct zbus_observer *obs, bool enabled)
+{
+	_ZBUS_ASSERT(obs != NULL, "obs is required");
+
+	obs->enabled = enabled;
+
+	return 0;
+}
+
+void __weak _zbus_notify_runtime_listeners(struct zbus_channel *chan)
+{
+}
+
+int __weak _zbus_notify_runtime_subscribers(struct zbus_channel *chan, uint64_t end_ticks)
+{
+	return 0;
+}
+
+void _zbus_notify_listeners(struct zbus_channel *chan)
+{
+	for (struct zbus_observer **obs = chan->observers; *obs != NULL; ++obs) {
+		if ((*obs)->enabled && ((*obs)->callback != NULL)) {
+			(*obs)->callback(chan);
+		}
+	}
+	_zbus_notify_runtime_listeners(chan);
+}
+
+int _zbus_notify_subscribers(struct zbus_channel *chan, uint64_t end_ticks)
+{
+	int err = 0;
+
+	for (struct zbus_observer **obs = chan->observers; *obs != NULL; ++obs) {
+		if ((*obs)->enabled && ((*obs)->queue != NULL)) {
+			err = k_msgq_put((*obs)->queue, &chan, _zbus_timeout_remainder(end_ticks));
+			if (err) {
+				return err;
+			}
+		}
+	}
+	return _zbus_notify_runtime_subscribers(chan, end_ticks);
+}
+
+int zbus_chan_pub(struct zbus_channel *chan, void *msg, k_timeout_t timeout)
+{
+	int err;
+	uint64_t end_ticks = sys_clock_timeout_end_calc(timeout);
+
+	_ZBUS_ASSERT(!k_is_in_isr(), "ISR blocked");
+	_ZBUS_ASSERT(chan != NULL, "chan is required");
+	_ZBUS_ASSERT(msg != NULL, "msg is required");
+
+	if (chan->read_only != 0) {
+		return -EPERM;
+	}
+
+	if (chan->validator != NULL && !chan->validator(msg, chan->message_size)) {
+		return -ENOMSG;
+	}
+
+	err = k_mutex_lock(chan->mutex, _zbus_timeout_remainder(end_ticks));
```

Comment:
The first blocking call can actually use the `timeout` variable directly, no need to re-calculate here.